crypto/cipher_extra/e_tls.c

```diff
@@ -45,19 +45,12 @@ typedef struct {
 OPENSSL_STATIC_ASSERT(EVP_MAX_MD_SIZE < 256,
                       mac_key_len_does_not_fit_in_uint8_t)
 
-OPENSSL_STATIC_ASSERT(sizeof(((EVP_AEAD_CTX *)NULL)->state) >=
-                          sizeof(AEAD_TLS_CTX),
-                      AEAD_state_is_too_small)
-#if defined(__GNUC__) || defined(__clang__)
-OPENSSL_STATIC_ASSERT(alignof(union evp_aead_ctx_st_state) >=
-                          alignof(AEAD_TLS_CTX),
-                      AEAD_state_has_insufficient_alignment)
-#endif
-
 static void aead_tls_cleanup(EVP_AEAD_CTX *ctx) {
-  AEAD_TLS_CTX *tls_ctx = (AEAD_TLS_CTX *)&ctx->state;
+  AEAD_TLS_CTX *tls_ctx = (AEAD_TLS_CTX *)ctx->state.ptr;
```

Comment:
I am probably missing something obvious. Why is this change needed? why allocate memory when I think the opaque array in `evp_aead_ctx_st_state` was supposed to store what is being allocated here?

Reply:
The existing code reuses the space in `EVP_AEAD_CTX->state` but that was never really an intended use for it. (It's designed to have enough space for AES-GCM.) Previously this worked because the `AEAD_TLS_CTX` was relatively small (because it contained an `HMAC_CTX` which then had pointers to other objects). Now that we are flattening `HMAC_CTX` (by removing pointers) and making it larger, we've done the same to `AEAD_TLS_CTX` and it no longer fits in `EVP_AEAD_CTX->state`. We *could* make `EVP_AEAD_CTX->state` larger, but that means that all uses of AES-GCM will spend extra memory when it is rarely needed. Instead, we just store a pointer in `EVP_AEAD_CTX->state` to newly allocated memory.

(Remember, the old code still allocates memory implicitly because the old `HMAC_CTX` contains pointers which require memory allocation.)